### PR TITLE
Rewrote parser to a more flexible version. Fixes #14 and #15

### LINKED ISSUE #14: Parser should interpret adjacent parentheses as multiplication
The parser interprets the following incorrectly:

```js
var exp = algebra.parse("(x + 2)(x + 2)");
exp.toString(); // x + 2
```

Currently, you have to do it like this:

```js
var exp = algebra.parse("(x + 2) * (x + 2)");
exp.toString(); // x^2 + 4x + 4
```

But it should be able to parse the top example too. @timrach can you take a look?

### PR DESCRIPTION
**Pull request by timrach**
The parser is not longer based on the shunting yard algorithm as it was not flexible enough. The new version is based on a grammar and therefore can be extended by any feature that can be expressed in the grammar. 
With the grammar it should be easy to include parsing future features. 
This commit also fixes issues 14 and 15.

### CODE REVIEW

**Comment by kevinb7** [comment]
Cool.  I'm happy for the switch to a recursive descent parser... mainly b/c I'm more familiar with them which will make it easier for me to help out.

File: test/parser-spec.js [hunk 1]
```diff
@@ -63,6 +65,19 @@ describe("Input validity", function() {
...
     });
 
+    it("should treat negative numbers correctly", function(){
+        var input = "x = -4";
+        var eqn = new Equation(new Expression("x"),new Expression(-4));
+        expect(algebra.parse(input)).toEqual(eqn);
+    });
+
+    it("should consecutive expressions as multiplication", function(){
+        var input = "(x + 2) * (x + 2)";
```
Comment (kevinb7) [inline test/parser-spec.js]: Shouldn't this test case be `(x + 2)(x + 2)`?

Comment (timrach) [inline test/parser-spec.js]: You are right, thanks. I fixed it.

**Comment by kevinb7** [comment]
Does it handle `5(x+2)` and `(x+2)5`?

**Comment by timrach** [comment]
It does handle 5(x+2) and (x+2)5. I added tests for that.

**Comment by kevinb7** [comment]
Awesome, thank you!

**Comment by nicolewhite** [comment]
Nice! I'll pull this in.

**Pull request merged**